client/beefy/rpc/src/lib.rs

```diff
@@ -141,6 +234,59 @@ mod tests {
 		(meta, rx)
 	}
 
+	#[test]
+	fn uninitialized_rpc_handler() {
+		let (io, _) = setup_io_handler();
+
+		let request = r#"{"jsonrpc":"2.0","method":"beefy_getFinalizedHead","params":[],"id":1}"#;
+		let response = r#"{"jsonrpc":"2.0","error":{"code":1,"message":"BEEFY RPC endpoint not ready"},"id":1}"#;
+
+		let meta = sc_rpc::Metadata::default();
+		assert_eq!(Some(response.into()), io.handle_request_sync(request, meta));
+	}
+
+	#[test]
+	fn latest_finalized_rpc() {
+		let (sender, stream) = BeefyBestBlockStream::<Block>::channel();
+		let (io, _) = setup_io_handler_with_best_block_stream(stream);
+
+		let header = Header {
+			parent_hash: BlakeTwo256::hash(b"1"),
+			number: 2,
+			state_root: BlakeTwo256::hash(b"3"),
+			extrinsics_root: BlakeTwo256::hash(b"4"),
+			digest: Digest { logs: vec![] },
+		};
+		sender.notify(|| Ok(header)).unwrap();
+
+		// Verify RPC `beefy_getFinalizedHead` returns expected header.
+		let request = r#"{"jsonrpc":"2.0","method":"beefy_getFinalizedHead","params":[],"id":1}"#;
+		let expected = "{\
+			\"jsonrpc\":\"2.0\",\
+			\"result\":{\
+				\"digest\":{\"logs\":[]},\
+				\"extrinsicsRoot\":\"0xeb8649214997574e20c464388a172420d25403682bbbb80c496831c8cc1f8f0d\",\
+				\"number\":\"0x2\",\
+				\"parentHash\":\"0x92cdf578c47085a5992256f0dcf97d0b19f1f1c9de4d5fe30c3ace6191b6e5db\",\
+				\"stateRoot\":\"0x581348337b0f3e148620173daaa5f94d00d881705dcbf0aa83efdaba61d2ede1\"\
+			},\
+			\"id\":1\
+		}";
+
+		let deadline = std::time::Instant::now() + std::time::Duration::from_secs(2);
+		while std::time::Instant::now() < deadline {
+			let meta = sc_rpc::Metadata::default();
+			if Some(expected.into()) == io.handle_request_sync(request, meta) {
+				// Success
+				return
+			}
+			std::thread::sleep(std::time::Duration::from_millis(50));
+		}
```

Comment:
I don't really like this but don't have any better suggestion right now.

Reply:
I improved the test a bit. To get rid of the deadline try-loop we'd need another comms channel between the tasks which is overkill imho. The current version of the test provides explicit results in all situations.